### ISSUE TITLE
[Epic] Helpers refactor: config, paths, smoke tests

### ISSUE THREAD

**Issue by wikefjol** [wikefjol/taxonml#25]
# Epic: Centralized Config & Path Handling (Prep-first), plus CI Smoke Tests

**Goal:** Establish a single, deterministic, easy-to-test path & config ingestion layer — starting with the **experiment prep** entry point — and set the groundwork to roll the same pattern into **pretrain** and **finetune** next.

---

## Scope & Sub-Issues

This epic **tracks** and **unblocks** the following issues:

* ✅ #3 `[core.logging] setup_logging()` — already done; not in scope beyond reuse
* ▶️ #5 `[core.io] New: read_yaml / write_json / clear_dir` *(we will implement read/write; no dir nuking in prep)*
* ▶️ #6 `[core.paths] New: central path rendering & run dirs` *(start with prep; pretrain/finetune later)*
* ▶️ #7 `[core.config] New: load_config(mode) → flat runtime dict` *(start with mode='prep')*
* ▶️ #22 `[CI/Tests] Smoke tests for load_config + path rendering`
* (Next pass) #20 `[Cleanup] Remove duplicated helpers from entry points`
* (Next pass) #19, #16 *(pipelines for prep; thin wrappers)*

---

### A. Config philosophy

* **No env variables** in runtime resolution. Inputs must be provided directly via YAML (absolute paths) or derived in code.
* **One loader, three modes:** `load_config(mode='prep'|'pretrain'|'finetune')`. This sprint implements **prep** only; others follow.
* **Return a flat runtime dict** (JSON-serializable) with fully resolved `Path` values under `cfg["paths"]`.
* **Fail fast** on missing keys / inconsistent values.

### B. Paths philosophy

* All path derivation lives in **`core.paths`**. Directory schema fragments remain **private** to this module.
* Entry points receive **absolute paths only**; they do not build path strings ad hoc.
* **No directory nuking** in prep. We **overwrite files** deterministically.
* Prep will **always produce both profiles**: `full/` and `debug/`.

  * `debug/` is produced using the debug subset rules; if the subset is empty, we still emit the profile folder with a clear summary (`rows: 0`) and warnings.
* Before writing **any** prep outputs, we **prompt for confirmation** if target files already exist (to avoid accidental overwrite).

  * Non-interactive bypass: `--yes` flag (alias `--force`).
  * We **do not** prompt in pretrain/finetune.

### C. Constants vs Config

* Column names & similar “wire protocol” strings live in **module-local constants**, *not* in YAML:

  * e.g., in `data.folds` / `labels.masks`: `FOLD_COL_EXP1 = "fold_exp1"`, `FOLD_COL_EXP2 = "fold_exp2"`.
* Directory schema fragments remain **private to `core.paths`**.
* YAML holds **experiment-level knobs and absolute roots**; everything else is derived.

### D. Logging & IO

* Use `core.logging.setup_logging(..., buffer_early=True)` and `attach_file_logger(...)` pattern (already landed).
* IO uses **atomic writes**: `core.io.write_json(..., atomic=True)` by default.

---

## Minimal YAML schema (prep-relevant)

The YAML **may** contain pretrain/finetune sections already, but `load_config(mode='prep')` will **only** read these keys:

```yaml
experiment:
  name: "sequence_fold_full"
  seed: 42
  union: "sequence_fold_full"   # key into data.inputs

data:
  experiments_root: "/abs/path/to/experiments"     # no env expansion
  input_dir: "/abs/path/to/inputs"                 # absolute path
  filenames:
    prep_summary: "prep_summary.json"
  inputs:
    sequence_fold_full: ["blast_union_A", "blast_union_B"]  # CSV basenames (no .csv)

label_space:
  levels: ["phylum","class","order","family","genus","species"]

prep:
  uppercase_sequences: true
  min_species_resolution: 0
  dedupe:
    scope: "within_species"
    keep_policy: "highest_resolution"
  folds:
    k: 10
    stratify_by: "species"
    seed: 42
  folds_species_group:
    k: 10
    stratify_by: "genus"
    seed: 42
  debug_subset:        # always executed; we always emit both profiles
    n_genera: 30
    min_species_per_genus: 5
    target_size: 20000
    seed: 42
```

> ✅ **Note:** `debug_subset.enabled` is no longer needed — prep **always** tries to build both profiles.

---

### `core.config.load_config(mode='prep') -> dict`

Returns a flat runtime dict with at least:

```python
{
  "experiment": {"name": str, "seed": int, "union": str},
  "data": {
    "experiments_root": Path,
    "input_dir": Path,
    "inputs": dict[str, list[str]],
    "filenames": {"prep_summary": str}
  },
  "label_space": {"levels": list[str]},
  "prep": {
    "uppercase_sequences": bool,
    "min_species_resolution": int,
    "dedupe": {"scope": str, "keep_policy": str},
    "folds": {"k": int, "stratify_by": str, "seed": int},
    "folds_species_group": {"k": int, "stratify_by": str, "seed": int},
    "debug_subset": {"n_genera": int, "min_species_per_genus": int, "target_size": int, "seed": int}
  },
  "paths": {
    "exp_root": Path,
    "logs_dir": Path,
    "prep_log": Path,
    "data_root": Path,                # experiments/<name>/data
    "profile_dirs": {
      "full": Path,                   # experiments/<name>/data/full
      "debug": Path                   # experiments/<name>/data/debug
    },
    "prepared_csv": {                 # handy aliases
      "full": Path,                   # .../full/prepared.csv
      "debug": Path                   # .../debug/prepared.csv
    },
    "label_space_json": {
      "full": Path,                   # .../full/label_space.json
      "debug": Path
    },
    "fold_masks": {
      "exp1": {"full": Path, "debug": Path},
      "exp2": {"full": Path, "debug": Path}
    },
    "prep_summary_json": Path         # .../data/prep_summary.json
  }
}
```

### `core.paths.prep_paths(experiments_root: Path, experiment_name: str) -> dict`

Builds the `paths` object above; schema fragments remain private to this module.

### `core.io`

* `read_yaml(Path) -> dict`
* `write_json(Path, obj: Any, *, atomic: bool = True) -> None`

---

### Phase 0 — Scaffolding & contracts

* [x] Create files:

  * `src/taxml/core/io.py`
  * `src/taxml/core/paths.py`
  * `src/taxml/core/config.py`
  * (constants live where used, no new global file needed for this sprint)
* [ ] Add unit test skeletons for `core.io`, `core.paths`, `core.config`.

### Phase 1 — `core.io`

* [x] Implement `read_yaml(path)`: strict loader, nice errors.
* [x] Implement `write_json(path, obj, atomic=True)`: write to `path.tmp` + `os.replace`.
* [ ] Tiny tests for round-trip and atomicity.

### Phase 2 — `core.paths` (prep only)

* [x] Implement `prep_paths(experiments_root, experiment_name)`:

  * Return structure under `paths` (see API above).
  * Create parent dirs where appropriate (`logs_dir`, `data_root`, profile dirs).
  * **Do not** create output files.
* [ ] Tests: ensure correct rendering for `full/` and `debug/`, and that paths are absolute.

### Phase 3 — `core.config.load_config('prep')`

* [x] Validate required keys (see YAML schema).
* [x] Resolve absolute `Path`s; **no env expansion**.
* [x] Inject `paths = core.paths.prep_paths(...)`.
* [x] Return a **flat dict** (not a class).
* [ ] Tests: cover happy path + missing key failures.

### Phase 4 — Wire `cli/exp_setup.py` to the loader

* [x] Replace inline YAML/env/path logic with `load_config('prep')`.
* [x] Early alias: `paths = cfg["paths"]` to keep call sites readable.
* [x] Implement **overwrite confirmation**:

  * If **any** of the known outputs exist (`prepared.csv`, `label_space.json`, `fold_masks_*`, `summary.json`, `prep_summary.json`):

    * Prompt: `About to overwrite existing prep artifacts under {paths['data_root']}. Continue? [yes/NO]`
    * Proceed only if **`yes`** (case-insensitive).
    * Add `--yes`/`--force` flag to bypass prompt (used in CI).
* [x] Always produce **both** profiles (`full`, `debug`). If debug subset is empty, still emit an empty profile with `summary.json` and warnings.
* [x] Keep file-level overwrite semantics; no directory deletion.
* [ ] Tests (integration-ish):

  * Create tiny synthetic CSVs.
  * Run prep with `--yes`; assert both profiles exist and files were written.
  * Run prep again; ensure it prompts; with `--yes` it overwrites.

### Phase 5 — CI smoke tests (#22)

* [ ] Add `pytest` jobs to CI:

  * `test_config_prep_loads_and_paths`: assert dict shape & absolute paths.
  * `test_prep_writes_both_profiles_and_masks`: on tiny data; assert files; assert masks length correctness vs label space.
  * `test_prep_overwrite_prompt_respected`: simulate prompt vs `--yes`.

> **Note:** Pretrain/finetune adoption will be a separate follow-up PR chain that uses the same loader & paths and then enables cleanup (#20).

---

## Deliverables / DoD

* `core.io`, `core.paths`, `core.config` implemented and tested for **prep**.
* `cli/exp_setup.py` uses `load_config('prep')` + `paths` object.
* Prep always writes **both** `full/` and `debug/` (even if debug is empty).
* Overwrite behavior:

  * Prompts by default if outputs exist.
  * Respects `--yes` to bypass (non-interactive safe).
* CI tests for loader, path rendering, and prep behaviors are green.

---

## Migration Notes / Breaking Changes

* **Removed** reliance on environment variables for paths in prep. YAML must specify absolute roots.
* `debug_subset.enabled` is **ignored**; debug is always attempted. Keep the numeric knobs (`n_genera`, etc.).
* Any existing ad-hoc path-building in `exp_setup.py` is replaced by `core.paths.prep_paths`.

---

## Risks & Mitigations

* **Accidental overwrite:** mitigated by interactive prompt and `--yes` gating in CI/automation.
* **Path drift across entry points:** mitigated by centralizing in `core.paths` and using it consistently later for pretrain/finetune.
* **Debug subset empty:** we still emit a profile with `rows: 0` and warnings to keep downstream tooling stable.

---

## Out-of-Scope (next PRs)

* Adopt `load_config('pretrain'|'finetune')` in those entry points (#17, #18).
* Move orchestration to `pipelines/prep.py` and keep CLI thin (#16, #19).
* Helper consolidation/cleanup across entry points (#20).
* Docs pass (#21) once helpers are stable.

---

## Example (what callers see)

```python
from taxml.core.config import load_config

cfg = load_config(mode="prep")
paths = cfg["paths"]  # single alias for readability

# Read inputs
input_dir = cfg["data"]["input_dir"]
union = cfg["experiment"]["union"]
members = cfg["data"]["inputs"][union]

# Outputs
full_csv  = paths["prepared_csv"]["full"]
debug_csv = paths["prepared_csv"]["debug"]
prep_log  = paths["prep_log"]
```

---

## Acceptance Checklist (maintainer)

* [ ] `core.io`, `core.paths`, `core.config` shipped with tests.
* [ ] `exp_setup.py` uses the new loader and prompts on overwrite.
* [ ] Both profiles always written; empty debug handled gracefully.
* [ ] CI pipeline green; `pytest -q` passes locally.
* [ ] Short README snippet added to explain `--yes` behavior.

---

**Comment by wikefjol** [comment]
Changed the yaml criterion for prep: 


```
experiment:
  name: "sequence_fold_full"
  seed: 42
  union: "sequence_fold_full"   # key into data.inputs

data:
  experiments_root: "/abs/path/to/experiments"     # no env expansion
  input_dir: "/abs/path/to/inputs"                 # absolute path
  filenames:
    prep_summary: "prep_summary.json"
  inputs:
    sequence_fold_full: ["blast_union_A", "blast_union_B"]  # CSV basenames (no .csv)

label_space:
  levels: ["phylum","class","order","family","genus","species"]

prep:
  uppercase_sequences: true
  min_species_resolution: 0
  dedupe:
    scope: "within_species"
    keep_policy: "highest_resolution"
  folds:
    k: 10
    stratify_by: "species"
    seed: 42
  folds_species_group:
    k: 10
    stratify_by: "genus"
    seed: 42
  debug_subset:        # always executed; we always emit both profiles
    n_genera: 30
    min_species_per_genus: 5
    target_size: 20000
    seed: 42

```
